File: Published/SourceCode/Java/SeparateChainingHashMap.java

import java.util.Iterator;

class SeparateChainingHashMap<K, V> implements Map<K, V> {
    Map<K, V>[] internalTable;
    int mapSize;

    SeparateChainingHashMap() {
        initialiseTable(MinCapacity);
    }

    @SuppressWarnings("unchecked")
    private void initialiseTable(int capacity) {
        internalTable = (Map<K, V>[]) new LinkedListMap[capacity];
        mapSize = 0;
    }

    static int MinCapacity = 8;
    static double MinLoadFactor = 0.5;
    static double MaxLoadFactor = 2.0;
    static double CapacityMultiplier = 1.5;

    private int hash(K key) {
        return (key.hashCode() & 0x7fffffff) % internalTable.length;
    }

    public V put(K key, V value) {
        int i = hash(key);
        Map<K, V> bin = internalTable[i];
        if (bin == null)
            bin = internalTable[i] = new LinkedListMap<>();
        V old = bin.put(key, value);
        if (old == null) {
            mapSize++;
            if (loadFactor() > MaxLoadFactor)
                resizeTable((int) (internalTable.length * CapacityMultiplier));
        }
        return old;
    }

    public V get(K key) {
        int i = hash(key);
        Map<K, V> bin = internalTable[i];
        return bin == null ? null : bin.get(key);
    }        

    public V remove(K key) {
        int i = hash(key);
        Map<K, V> bin = internalTable[i];
        if (bin == null)
            return null;
        V removed = bin.remove(key);
        if (removed != null) {
            mapSize--;
            if (loadFactor() < MinLoadFactor) 
                resizeTable((int) (internalTable.length / CapacityMultiplier));
        }
        return removed;
    }            

    public boolean containsKey(K key) {
        int i = hash(key);
        Map<K, V> bin = internalTable[i];
        return bin != null && bin.containsKey(key);
    }

    private void resizeTable(int newCapacity) {
        if (newCapacity < MinCapacity) return;
        Map<K, V>[] oldTable = internalTable;
        initialiseTable(newCapacity);
        for (Map<K, V> bin : oldTable)
            if (bin != null)
                for (K k : bin)
                    put(k, bin.get(k));
    }

    public boolean isEmpty() {
        return mapSize == 0;
    }

    public int size() {
        return mapSize;
    }

    public double loadFactor() {
        return (double) mapSize / internalTable.length;
    }

    public Iterator<K> iterator() {
        return new HashMapIterator();
    }

    private class HashMapIterator implements Iterator<K> {
        private int bucketIndex;
        private Iterator<K> bucketIterator;
        HashMapIterator() {
            bucketIndex = 0;
            bucketIterator = internalTable[bucketIndex].iterator();
        }
        public boolean hasNext() {
            while (!bucketIterator.hasNext()) {
                bucketIndex++;
                if (bucketIndex >= internalTable.length) return false;
                bucketIterator = internalTable[bucketIndex].iterator();
            }
            return true;
        }
        public K next() {
            hasNext();
            return bucketIterator.next();
        }
    }


/***************************************************************************************/
/** What comes below is purely for debugging and testing purposes - it can be removed **/

    public void _printMap() {
        System.out.print(internalTable.length + " [ ");
        for (int i=0; i<internalTable.length; i++) {
            if (i > 0) System.out.print("| ");
            Map<K, V> bin = internalTable[i];
            if (bin == null) System.out.print("? ");
            else if (bin.isEmpty()) System.out.print("- ");
            else for (K k : bin) System.out.print(k + ":" + get(k) + " ");
        }
        System.out.println("] " + size() + " {" + loadFactor() + "}");
    }

    public static void main(String[] args) {
        SeparateChainingHashMap<String, Integer> map = new SeparateChainingHashMap<>();
        map._printMap();
        System.out.println("Putting values");
        for (int i=0; i<40; i++) {
            map.put(String.valueOf((char)((i%25)+65)), i);
            if (i % 5 == 0) map._printMap();
        }
        map._printMap();
        System.out.println("Removing values");
        int i = 7;
        while (!map.isEmpty()) {
            map.remove(String.valueOf((char)((i%25)+65)));
            i += 11;
            if (map.size() % 5 == 0) map._printMap();
        }
    }
}

File: Published/SourceCode/Python/API/SeparateChainingHashMap.py

from API import Map
from LinkedMap import LinkedMap

class SeparateChainingHashMap(Map):
    _minCapacity = 8
    _minLoadFactor = 0.5
    _maxLoadFactor = 2.0
    _capacityMultiplier = 1.5

    def __init__(self):
        self._initialiseTable(self._minCapacity)

    def _initialiseTable(self, capacity):
        capacity = int(capacity)
        self._internalTable = [None] * capacity
        for i in range(capacity):
            self._internalTable[i] = LinkedMap()
        self._mapSize = 0

    def _hash(self, key):
        return (hash(key) & 0x7fffffff) % len(self._internalTable)

    def put(self, key, value):
        i = self._hash(key)
        old = self._internalTable[i].put(key, value)
        if old is None:
            self._mapSize += 1
            if self.loadFactor() > self._maxLoadFactor:
                self._resizeTable(len(self._internalTable) * self._capacityMultiplier)
        return old

    def get(self, key):
        i = self._hash(key)
        return self._internalTable[i].get(key)

    def remove(self, key):
        i = self._hash(key)
        removed = self._internalTable[i].remove(key)
        if removed is not None:
            self._mapSize -= 1
            if self.loadFactor() < self._minLoadFactor:
                self._resizeTable(len(self._internalTable) / self._capacityMultiplier)
        return removed

    def containsKey(self, key):
        i = self._hash(key)
        return self._internalTable[i].containsKey(key)

    def _resizeTable(self, newCapacity):
        if newCapacity < self._minCapacity: return
        oldTable = self._internalTable
        self._initialiseTable(int(newCapacity))
        for bin in oldTable:
            for k in bin:
                self.put(k, bin.get(k))

    def isEmpty(self):
        return self._mapSize == 0

    def size(self):
        return self._mapSize

    def loadFactor(self):
        return self._mapSize / len(self._internalTable)

    def __iter__(self):
        for bin in self._internalTable:
            for key in bin:
                yield key


#######################################################################################
## What comes below is purely for debugging and testing purposes - it can be removed ##

def _printMap(map):
    print(len(map._internalTable), "[",
          " | ".join(" ".join(f"{k}:{bin.get(k)}" for k in bin) for bin in map._internalTable),
          "]", m.size(), "{", m.loadFactor(), "}")

if __name__ == '__main__':
    m = SeparateChainingHashMap()
    _printMap(m)
    print("Putting values")
    for i in range(40):
        m.put(chr((i%25)+65), i)
        if i % 5 == 0: _printMap(m)
    _printMap(m)
    print("Removing values")
    i = 7
    while not m.isEmpty():
        m.remove(chr((i%25)+65))
        i += 11
        if m.size() % 5 == 0: _printMap(m)
